File: frontend/src/lib/api/workspaces.ts
import { apiDelete, apiGet, apiPatch, apiPost } from './client';
import type { Page } from './chatbot';

export type WorkspaceRole = 'owner' | 'admin' | 'member';

export interface Workspace {
  id: string;
  name: string;
  description: string | null;
  created_by_user_id: string;
  created_at: string;
  updated_at: string;
}

export interface WorkspaceMember {
  id: string;
  workspace_id: string;
  user_id: string;
  workspace_role: WorkspaceRole;
  created_at: string;
  updated_at: string;
}

export interface WorkspaceCreateInput {
  name: string;
  description?: string | null;
}

export interface WorkspaceUpdateInput {
  name?: string;
  description?: string | null;
}

export function listWorkspaces(page = 1, pageSize = 20): Promise<Page<Workspace>> {
  return apiGet<Page<Workspace>>(`/api/workspaces?page=${page}&page_size=${pageSize}`);
}

export function getWorkspace(id: string): Promise<Workspace> {
  return apiGet<Workspace>(`/api/workspaces/${id}`);
}

export function createWorkspace(input: WorkspaceCreateInput): Promise<Workspace> {
  return apiPost<Workspace>('/api/workspaces', input);
}

export function updateWorkspace(id: string, input: WorkspaceUpdateInput): Promise<Workspace> {
  return apiPatch<Workspace>(`/api/workspaces/${id}`, input);
}

export function listWorkspaceMembers(workspaceId: string): Promise<{ items: WorkspaceMember[] }> {
  return apiGet<{ items: WorkspaceMember[] }>(`/api/workspaces/${workspaceId}/members`);
}

export function addWorkspaceMember(
  workspaceId: string,
  userId: string,
  workspaceRole: WorkspaceRole,
): Promise<WorkspaceMember> {
  return apiPost<WorkspaceMember>(`/api/workspaces/${workspaceId}/members`, {
    user_id: userId,
    workspace_role: workspaceRole,
  });
}

export function changeWorkspaceMemberRole(
  workspaceId: string,
  userId: string,
  workspaceRole: WorkspaceRole,
): Promise<WorkspaceMember> {
  return apiPatch<WorkspaceMember>(`/api/workspaces/${workspaceId}/members/${userId}`, {
    workspace_role: workspaceRole,
  });
}

export function removeWorkspaceMember(workspaceId: string, userId: string): Promise<void> {
  return apiDelete<void>(`/api/workspaces/${workspaceId}/members/${userId}`);
}

File: backend/schemas/workspaces.py
"""Request/response models for workspaces and workspace membership."""
from typing import List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from backend.schemas.common import Page, UtcDatetime

WorkspaceRole = Literal["owner", "admin", "member"]


class WorkspaceCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200, examples=["Acme Corp Security"])
    description: Optional[str] = Field(default=None, max_length=2000)


class WorkspaceUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=2000)


class WorkspaceItem(BaseModel):
    id: UUID = Field(examples=["7c9e6679-7425-40de-944b-e07fc1f90ae7"])
    name: str
    description: Optional[str] = None
    created_by_user_id: UUID
    created_at: UtcDatetime
    updated_at: UtcDatetime


class WorkspacePage(Page[WorkspaceItem]):
    pass


class WorkspaceMemberAdd(BaseModel):
    user_id: UUID
    workspace_role: WorkspaceRole


class WorkspaceMemberRoleChange(BaseModel):
    workspace_role: WorkspaceRole


class WorkspaceMemberItem(BaseModel):
    id: UUID
    workspace_id: UUID
    user_id: UUID
    workspace_role: WorkspaceRole
    created_at: UtcDatetime
    updated_at: UtcDatetime


class WorkspaceMemberList(BaseModel):
    items: List[WorkspaceMemberItem]
